### PR TITLE
Deprecation: 	CRM-20572 Remove jcalendar from Contrib Sybunt custom Search

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
* [CRM-20572: Remove jcalendar from Contrib Sybunt custom Search](https://issues.civicrm.org/jira/browse/CRM-20572)

### CODE REVIEW

**Comment by seamuslee001** [comment]
I've given this a bit of a check around and can't seem to see anything broken, @eileenmcnaughton how would any smart groups created based off this search work?

**Comment by eileenmcnaughton** [comment]
Oh smart groups ... hmm. I'm inclined to think this is obscure enough it might be ok not to worry too much about that. I'm pretty sure the values would have been formatted before saving?

**Comment by seamuslee001** [comment]
I would say so but just thought i should chuck the curve ball out there,

**Comment by seamuslee001** [comment]
@eileenmcnaughton i ran into an error when trying to load the smart group that i saved based  on the old code
```
    [to_string] => [db_error: message="DB Error: unknown error" code=-1 mode=callback callback=CRM_Core_Error::handle prefix="" info="CREATE TEMPORARY TABLE CustomSearch_SYBNT_temp AS (
SELECT
sum(contrib_1.total_amount) AS donation_amount,
count(contrib_1.id) AS donation_count
, contact_a.id, display_name
FROM       civicrm_contact AS contact_a
LEFT JOIN  civicrm_contribution contrib_1 ON contrib_1.contact_id = contact_a.id

WHERE      contrib_1.contact_id = contact_a.id
AND        contrib_1.is_test = 0
            AND contrib_1.receive_date >= '07/01/2016' AND contrib_1.receive_date <=  '12/31/2017' AND  contact_a.is_deleted = 0
GROUP BY   contact_a.id

ORDER BY   donation_amount desc
) [nativecode=1292 ** Incorrect datetime value: '07/01/2016' for column 'receive_date' at row 1]"]
)
```

**Comment by colemanw** [comment]
It would be useful to have a generic way to update smart groups since we keep running into this sort of thing.

**Comment by monishdeb** [comment]
I'll take a look at this remaining issue of smart group and will make necessary fix. As I recall, did similar fix to fetch n set date values from saved search for a custom date field.

**Comment by eileenmcnaughton** [comment]
ohh you rock @monishdeb !

**Comment by monishdeb** [comment]
@eileenmcnaughton @colemanw here's what I have done to fix the old saved data of a smart group which is affecting this custom search. Issue was the old date values are stored in ```d/m/y``` format and the expected datepicker format is ```Y-m-d``` (which is also acceptable MySQL date value). So in order to format the old to new datepicker format, while decoding it will lookout for special function ```formatSavedSearchFields(...)``` in the custom class [here](https://github.com/civicrm/civicrm-core/pull/10355/files#diff-54576153cb6a72d6ea7e22a5e8ca63f2R191), that format saved-search values for selected fields as here for SYBNT search form, those were [date fields](https://github.com/civicrm/civicrm-core/pull/10355/files#diff-b5b768f7ed02b402fbe39e1e3a14f553R410).

   So this special function can be used by other custom search form to resolved any of such oddities. Thoughts?

File: CRM/Contact/BAO/SavedSearch.php [hunk 1]
```diff
@@ -87,11 +87,13 @@ public static function retrieve(&$params, &$defaults) {
...
    * @param int $id
    *   The id of the saved search.
+   * @param bool $isCustomSearch
+   *   Are you fetching formValues for a custom search form?
    *
    * @return array
    *   the values of the posted saved search used as default values in various Search Form
    */
-  public static function getFormValues($id) {
+  public static function getFormValues($id, $isCustomSearch = FALSE) {
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/SavedSearch.php]: Thanks @monishdeb - just at the very trivial level - given all the other checks you have - do we need this parameter - it seems that the method_exists & the customSearchClass check should be enough

**Comment by monishdeb** [comment]
@eileenmcnaughton you are right, silly me :(

**Comment by eileenmcnaughton** [comment]
@seamuslee001  - does this resolve it now?

**Comment by monishdeb** [comment]
ping @seamuslee001 @eileenmcnaughton

File: CRM/Contact/Form/Search/Custom/ContribSYBNT.php [hunk 2]
```diff
@@ -43,7 +43,7 @@ class CRM_Contact_Form_Search_Custom_ContribSYBNT extends CRM_Contact_Form_Searc
    * @param $formValues
    */
   public function __construct(&$formValues) {
-    $this->_formValues = $formValues;
+    $this->_formValues = self::formatSavedSearchFields($formValues);
```
Comment (eileenmcnaughton) [inline CRM/Contact/Form/Search/Custom/ContribSYBNT.php]: works if you change it to     self::formatSavedSearchFields($formValues);
$this->_formValues = $formValues;

File: CRM/Core/BAO/ActionSchedule.php [hunk 3]
```diff
@@ -556,10 +556,7 @@ protected static function sendReminderSms($tokenRow, $schedule, $toContactID) {
       'provider_id' => $schedule->sms_provider_id,
       'activity_subject' => $messageSubject,
     );
-    $activityTypeID = CRM_Core_OptionGroup::getValue('activity_type',
-      'SMS',
-      'name'
-    );
+    $activityTypeID = CRM_Core_PseudoConstant::getKey('CRM_Activity_BAO_Activity', 'activity_type_id', 'SMS');
```
Comment (eileenmcnaughton) [inline CRM/Core/BAO/ActionSchedule.php]: probably better left out as unrelated & 'swept up'

**Comment by eileenmcnaughton** [comment]
@monishdeb I had a specific issue on this per comments. But, I also worry about the possibility of the site date being weird & wonderful & not being correctly translated by strtotime & if that is the case we need to 'know'. The only way I can think of to 'know' is to do it in the upgrade script - e.g add a function like 

```
  /**
   * Convert a saved search to use more recent date field formatting.
   *
   * When using datepicker dates are all saved as 2017-09-30 but the old field type
   * was variable, depending on site date configuration.
   *
   * @param $dateFields
   * @param $class
   */
  public static function convertSavedSearch($dateFields, $class) {
    $searchID = civicrm_api3('OptionValue', 'getvalue', array('label' => $class, 'return' => 'value'));
    $dao = CRM_Core_DAO::executeQuery('SELECT id, form_values FROM civicrm_saved_search WHERE custom_search_id = %1', array(1 => array($searchID, 'Integer')));
    while ($dao->fetch()) {
      $formValues = unserialize($dao->form_values);
      foreach($dateFields as $dateField) {
        if (isset($formValues[$dateField])) {
          // Note I just tapped this out - not sure it's the right function but need to assume it is in
          // site date format & use the function to convert from there.
          $formValues[$dateField] = date('Y-m-d', CRM_Utils_Date::format($formValues[$dateField]));
        }
      }
      $dao->form_values = serialize($formValues);
      $dao->save();
    }
  }
```
Note the above is not tested code.

**Comment by monishdeb** [comment]
@eileenmcnaughton earlier I thought about handling it in upgrade script but then I recall a similar issue where I need to fix values of check-box field >due-to> format mis-match >due-to> replacing the advance multi-select widget with select2. Then after discussion with dgg :) we decided to bring changes in code which is able to handle any oddities in format rather than handling it on upgrade too to reduce overhead. This was the PR  https://github.com/civicrm/civicrm-core/pull/7018/files

In this its just one custom search class so kept it simple. Or should I still include the upgrade code?

**Comment by eileenmcnaughton** [comment]
@monishdeb what worries me specifically about this that might not be the same with that one is that we can't identify on the fly whether it is a new date format or old one - which makes me swing towards upgrade.

Another disadvantage of 'just handling' is that we accumulate code in our codebase that might only be needed short term

File: CRM/Contact/Form/Search/Custom/ContribSYBNT.php (hunk 2 above)
Comment (monishdeb) [reply CRM/Contact/Form/Search/Custom/ContribSYBNT.php]: I thought so but I encountered an error at one place (can't recall the use-case). Should I still revert this change?

**Comment by monishdeb** [comment]
@eileenmcnaughton I have added the upgrade code, commit - https://github.com/civicrm/civicrm-core/pull/10355/commits/c80fb7c9b087f597c96072ee6afce47fb6cf52a4 Please have a look

**Comment by eileenmcnaughton** [comment]
@monishdeb stale again :-(. Let's aim to get it sorted at the end of the week after the rc is cut (since that seems to less risk staleness). Can you put the upgrade script rev to 4.7.27?

**Comment by monishdeb** [comment]
@eileenmcnaughton moved the upgrade code to 4.7.27

**Comment by seamuslee001** [comment]
@monishdeb @eileenmcnaughton can you rebase this please. On the whole the code works and i think we should merge this, I can test the upgrade when we get to RC for 4.7.27

**Comment by monishdeb** [comment]
@eileenmcnaughton @seamuslee001 I have rebased the PR

**Pull request merged**